docs/language/learn-ql/cpp/guards.rst

```diff
@@ -0,0 +1,53 @@
+Using the guards library in C and C++
+=====================================
+
+Overview
+--------
+The guards library (defined in ``semmle.code.cpp.controlflow.Guards``) provides a class ``GuardCondition`` representing Boolean values that are used to make control flow decisions. A ``GuardCondition`` is considered to guard a basic block if the block is only reached if the ``GuardCondition`` evaluated a certain way. For instance, in the following code, ``x < 10`` is a ``GuardCondition``, and it guards all the code before the return statement.
```

Comment:
My earlier comment was supposed to be anchored here, but GitHub made that hard to see. Here it is again.

> I think in my head I've been using guard (as a verb) as a synonym for control. But it's an edge that controls a block, whereas here you're talking about an expression guarding a block. Do you use guard as a synonym for dominate?

I just asked around, and the Copenhagen Consensus is that _controls_ and _guards_ are synonyms (so we all agree on that) and that we only say "Expression `e` guards block `b`" when it's implicit which edge out of `e` we're talking about (most often the true-edge). Furthermore, when an _edge_ dominates a block, that's the same as guarding it.

Reply:
So in the following sample, you'd say that the true edge out of the call to `isValid` controls the block starting with `performAction()`, and the false edge controls the block containing `logFailure`, but wouldn't say that the call to `isValid` controls anything?
```cpp
    if(!isValid(accessToken)) {
      logFailure();
      return 0;
    }
    performAction();
    return succeeded;
```

Reply:
The qldocs for the C++, C#, and Java versions of `GuardCondition.controls` are all phrased as the condition controlling blocks when evaluating a certain way, rather than edges controlling blocks. I'm not sure if that's because it was the preferred definition at the time or just because it was easier to phrase member predicate qldoc that way.

Reply:
> So in the following sample, you'd say that the true edge out of the call to `isValid` controls the block starting with `performAction()`, and the false edge controls the block containing `logFailure`, but wouldn't say that the call to `isValid` controls anything?

Yes. Except that the AST CFG doesn't skip `!`, so there are no true/false edges out of the call to `isValid`. There are just true/false edges out of the `!`.